datumaro/components/operations.py

```diff
@@ -106,40 +108,95 @@ def merge(cls, *sources):
-
-            if existing_item.image.path != current_item.image.path:
-                if not existing_item.image.path:
-                    image._path = current_item.image.path
-
-            if all([existing_item.image._size, current_item.image._size]):
-                if existing_item.image._size != current_item.image._size:
-                    raise MismatchingImageInfoError(
-                        (existing_item.id, existing_item.subset),
-                        existing_item.image._size, current_item.image._size)
-            elif existing_item.image._size:
-                image._size = existing_item.image._size
+                m_val = a_val
+
+            merged[name] = m_val
+
+        return merged
+
+    @staticmethod
+    def merge_images(item_a: DatasetItem, item_b: DatasetItem) -> Image:
+        image = None
+
+        if item_a.has_image and item_b.has_image:
+            if item_a.image.path and item_b.image.path and \
+                    item_a.image.path != item_b.image.path and \
+                    item_a.image.has_data is item_b.image.has_data:
+                # We use has_data as a replacement for path existence check
+                # - If only one image has data, we'll use it. The other
+                #   one is just a path metainfo, which is not significant
+                #   in this case.
+                # - If both images have data or both don't, we need
+                #   to compare paths.
+                #
+                # Different paths can aclually point to the same file,
+                # but it's not the case we'd like to allow here to be
+                # a "simple" merging strategy used for extractor joining
+                raise MismatchingImagePathError(
+                    (item_a.id, item_a.subset),
+                    item_a.image.path, item_b.image.path)
+
+            if item_a.image.has_size and item_b.image.has_size and \
+                    item_a.image.size != item_b.image.size:
+                raise MismatchingImageInfoError(
+                    (item_a.id, item_a.subset),
+                    item_a.image.size, item_b.image.size)
+
+            # Avoid direct comparison here for better performance
+            # If there are 2 "data-only" images, they won't be compared and
+            # we just use the first one
+            if item_a.image.has_data:
+                image = item_a.image
+            elif item_b.image.has_data:
+                image = item_b.image
+            elif item_a.image.path:
+                image = item_a.image
+            elif item_b.image.path:
+                image = item_b.image
+            elif item_a.image.has_size:
+                image = item_a.image
+            elif item_b.image.has_size:
+                image = item_b.image
             else:
-                image._size = current_item.image._size
-        elif existing_item.has_image:
-            image = existing_item.image
+                image = item_a.image
```

Comment:
Shouldn't this branch be impossible? If we got here, it means that neither image has a path, data, _or_ size, which should not be possible.

Reply:
Yes, currently it should be unreachable, but it will be a fallback if `Image` class is changed.

Reply:
Shouldn't it throw an exception, then? If this branch is reached due to `Image` changing, then you can't know which image should be selected.

Reply:
Yes, probably, it will be better.